src/operations/blas3/trsm.hpp

```diff
@@ -0,0 +1,158 @@
+/***************************************************************************
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ **************************************************************************/
+
+
+#ifndef SYCL_BLAS_BLAS3_REF_TRSM_HPP
+#define SYCL_BLAS_BLAS3_REF_TRSM_HPP
+
+#include "operations/blas3_trees.h"
+#include "views/view.h"
+
+#include <CL/sycl.hpp>
+
+namespace blas {
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+SYCL_BLAS_INLINE
+DiagonalBlocksInverter<UnitDiag, Upper, matrix_t>::DiagonalBlocksInverter(
+    matrix_t& A, matrix_t& invA)
+    : A_(A), invA_(invA), N_(A_.get_size_col()), lda_(A_.getSizeL()) {}
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+SYCL_BLAS_INLINE bool
+DiagonalBlocksInverter<UnitDiag, Upper, matrix_t>::valid_thread(
+    cl::sycl::nd_item<1> id) const {
+  return true;
+}
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+SYCL_BLAS_INLINE void DiagonalBlocksInverter<UnitDiag, Upper, matrix_t>::bind(
+    cl::sycl::handler& cgh) {
+  A_.bind(cgh);
+  invA_.bind(cgh);
+}
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+SYCL_BLAS_INLINE void DiagonalBlocksInverter<
+    UnitDiag, Upper, matrix_t>::adjust_access_displacement() {
+  A_.adjust_access_displacement();
+  invA_.adjust_access_displacement();
+}
+
+template <bool UnitDiag, bool Upper, typename matrix_t>
+template <typename local_memory_t>
+SYCL_BLAS_INLINE void DiagonalBlocksInverter<UnitDiag, Upper, matrix_t>::eval(
+    local_memory_t localMem, cl::sycl::nd_item<1> item) noexcept {
+  auto A = A_.get_data().get_pointer() + A_.get_access_displacement();
+  auto invA = invA_.get_data().get_pointer() + invA_.get_access_displacement();
+  value_t* local = localMem.localAcc.get_pointer();
+
+  const size_t threadIndex = item.get_local_id(0);
```

Comment:
Is there any reason you are using size_t here for index? for the rest of the BLAS we are using `int` as BLAS interface strictly saying that the index type should be `signed int`. Can you convert all the index type to be the template parameter  `indext_t` the one that has been passed by the user in calling TRSM and propagate it to this class instead of hardcode `size_t`. So, in this case, all the value type are `value_t` and the index types are `index_t` so you don't need to convert `for` loop index to have an arbitrary type and all will be `index_t` and no hardcoded type will be in the kernel. This will allow us to later switch the index type to whatever is required

Reply:
You can see what we have done in gemm. You can see it there : https://github.com/codeplaysoftware/sycl-blas/blob/master/src/operations/blas3/gemm_local.hpp#L270
You can cast it to signed if it gave you a warning

Reply:
The more important thing is the combination of signed and unsigned should not exist

Reply:
Everything was changed to `index_t`